ouroboros-network/src/Ouroboros/Network/Subscription/Common.hs

```diff
@@ -0,0 +1,383 @@
+                    activeCons  <- readTVar conThreads
+                    unless (null activeCons) retry
+
+                valencyLeft <- atomically $ readTVar valencyVar
+                if valencyLeft == 0
+                   then traceWith tracer SubscriptionTraceSubscriptionRunning
+                   else traceWith tracer SubscriptionTraceSubscriptionFailed
+
+
+    doConnect conThreads localAddr remoteAddr =
+        bracket
+            ( do
+                tid <- myThreadId
+                atomically $ do
+                    modifyTVar' conThreads (Set.insert tid)
+                    modifyTVar' threadPool (Set.insert tid)
+                sd <- Socket.socket (sockAddrFamily remoteAddr) Socket.Stream
+                        Socket.defaultProtocol
+                hasRefVar <- newTVarM False
+                return (sd, hasRefVar)
+            )
+            (\(sd,hasRefVar) -> do
+                traceWith tracer $ SubscriptionTraceConnectCleanup remoteAddr
+
+                tid <- myThreadId
+                atomically $ do
+                    modifyTVar' conThreads (Set.delete tid)
+                    modifyTVar' threadPool (Set.delete tid)
+                hasRef <- atomically $ readTVar hasRefVar
+                when hasRef $ do
+                    localAddr' <- Socket.getSocketName sd
+                    removeConnection tbl remoteAddr localAddr'
+
+                Socket.close sd
+
+            )
+            (\(sd,hasRefVar)  -> do
+                Socket.setSocketOption sd Socket.ReuseAddr 1
+#if !defined(mingw32_HOST_OS)
+                Socket.setSocketOption sd Socket.ReusePort 1
+#endif
+                Socket.bind sd localAddr
+                traceWith tracer $ SubscriptionTraceConnectStart remoteAddr
+                res <- try $ Socket.connect sd remoteAddr
+                case res of
+                        Left (e :: SomeException) -> do
+                            traceWith tracer $ SubscriptionTraceConnectException remoteAddr e
+                            throwM e
+                        Right _ -> return ()
+
+                localAddr' <- Socket.getSocketName sd
+
+                -- We successfully connected, increase valency and start the app
+                tid <- myThreadId
+                result <- atomically $ do
+                        v <- readTVar valencyVar
+                        if v > 0 then do
+                                    addConnection tbl remoteAddr localAddr' [valencyVar]
+                                    writeTVar hasRefVar True
+                                    modifyTVar' conThreads (Set.delete tid)
+                                    if v == 1 then return ConnectSuccessLast
+                                              else return ConnectSuccess
+                                else return ConnectFail
+                traceWith tracer $ SubscriptionTraceConnectEnd remoteAddr result
+                case result of
+                    ConnectSuccess -> k sd
```

Comment:
I wouldn't mind adding a todo, that this (and below) is the place where we can catch all the exceptions coming from the application.  It also could be added on top of `doConnect` since `bracket` and `onException` commute.